### ISSUE TITLE
SRP target occasionally not present after a reboot

### ISSUE THREAD

**Issue by dswartz** [astersmith/esos#109]
I've seen this a couple of times now.  I reboot the ESOS host, and when it comes up, and I login, the TUI says there is no SRP target.  Dropping to a shell, and doing 'lspci' shows the mellanox connectx-2 is there.  If I reboot again, voila, it's present.  I am speculating some kind of race condition?  If this happens again, do you want me to shoot you a support bundle?  Something else?

**Comment by msmith626** [comment]
Yeah, I'd need a support bundle generated when the issue occurs (reboot and no targets).

--Marc

**Comment by dswartz** [comment]
10-4.

**Comment by dswartz** [comment]
Happened on 1st boot with zfs.  I have tried rebooting several times and the target just does not show up.  And cannot be added dynamically.  Finally (out of desperation), I deleted /etc/scst.conf and rebooted, and it appeared.  It seems to be something about whatever auto-generates SRP targets.  It is there now, so I can make progress...

**Comment by msmith626** [comment]
I would suspect some type of an issue with the hardware... maybe re-seat the IB HCA card(s). Unfortunately the support bundle doesn't include everything... something to add in: Run the command 'dmesg' and see if the system detects the IB HCAs at all -- I'd suspect not.

--Marc

**Comment by dswartz** [comment]
On 2015-04-18 21:33, Marc A. Smith wrote:
> I would suspect some type of an issue with the hardware... maybe
> re-seat the IB HCA card(s). Unfortunately the support bundle doesn't
> include everything... something to add in: Run the command 'dmesg' and
> see if the system detects the IB HCAs at all -- I'd suspect not.

No, it is detected just fine.  Shows up as this:

02:00.0 InfiniBand: Mellanox Technologies MT26428 [ConnectX VPI PCIe 2.0 
5GT/s - IB QDR / 10GigE] (rev b0)
         Subsystem: Mellanox Technologies Device 0022
         Flags: bus master, fast devsel, latency 0, IRQ 17
         Memory at dfd00000 (64-bit, non-prefetchable) [size=1M]
         Memory at de000000 (64-bit, prefetchable) [size=8M]
         Capabilities: [40] Power Management version 3
         Capabilities: [48] Vital Product Data
         Capabilities: [9c] MSI-X: Enable+ Count=128 Masked-
         Capabilities: [60] Express Endpoint, MSI 00
         Capabilities: [100] Alternative Routing-ID Interpretation (ARI)
         Capabilities: [148] Device Serial Number 00-02-c9-03-00-4c-ed-9a
         Kernel driver in use: mlx4_core

(this is in the working config, btw...)

**Comment by msmith626** [comment]
Yeah, try 'dmesg' when its not working and see if it shows up.

**Comment by dswartz** [comment]
On 2015-04-18 21:38, Marc A. Smith wrote:
> Yeah, try 'dmesg' when its not working and see if it shows up.

I should have been clearer, sorry.  I did in fact do so when I was 
flailing around trying to get it to work (including multiple reboots...) 
  It seems to have something to do with the initial creation of 
/etc/scst.conf.  Like I said, I deleted it, and rebooted, and it 
suddenly shows up :)

**Comment by msmith626** [comment]
But on your first boot of ESOS, you wouldn't have an '/etc/scst.conf' file,
which would be the same as when you deleted that file and rebooted without
syncing. And it didn't work on the first boot when you didn't have an
scst.conf file either -- right?

**Comment by dswartz** [comment]
Sorry I missed this - thought I had replied.  I honestly can't remember the exact sequence of events.  I'm pretty sure it's failed the first time at least once.  Seems to be some kind of timing/race where the HW is not ready or something?  I'm pretty sure at least once I 'fixed' this by doing 'scstadmin --config xxx', where 'xxx' is the previous good config file, and lo and behold it was working...

**Comment by msmith626** [comment]
Definitely could be something with timing... the IB modules are loaded and
then a bit later in the init process SCST is loaded. Maybe try inserting a
"sleep 10" or similar in /etc/rc.d/rc.2 somewhere after the "early"
services block, but before the "late" services block.

--Marc

**Comment by dswartz** [comment]
I've had this happen with all 3 MHQH19B-XTR cards I got on ebay.  The cards *seem* to work okay, so it's hard to believe all 3 are bad in exactly the same way.  The fact that the hardware bring-up happens well before anything SCST related makes me suspect some kind of bug in the IB driver (or at least the parts relating to this particular HW).  Or possibly some oddness wrt the PCI-E slot in question.  In any event, doesn't seem ESOS is culpable.  Closing this...
